raft/server.cc

```diff
@@ -1282,6 +1288,13 @@ future<> server_impl::set_configuration(config_member_set c_new, seastar::abort_
     }
 
     _stats.add_config++;
+
+    if (_non_joint_conf_commit_promise) {
+        logger.warn("[{}] set_configuration: a configuration change is still in progress (at index: {}, config: {})",
+            _id, _fsm->log_last_conf_idx(), cfg);
+        throw conf_change_in_progress{};
```

Comment:
Why throw here? Can't we wait for the previous conf change to finish instead? Throwing like that gives the caller no chance to avoid an occasional throw - since it throws even after the conf change is finished in fsm but the promise is still not resolved because fsm output is not fully handled yet.

Reply:
A caller still did not get a notification about the config change completion, so it means it issues a new one without waiting for previous one to complete. In such case it has to be prepared to handle the exception.

Reply:
It doesn't have to be the same caller.

Reply:
> It doesn't have to be the same caller.

It does not matter. It issued a new config change before previous one returns. In this scenario such error is expected.

Reply:
A typical "guard" for issuing is get_configuration().is_joint(). It used to be a reliable way to find out if you can append a new config. It is no longer true.

Reply:
That's somewhat true, but only if you do it on a leader. But even then in a presence of concurrent configuration changes even if you will wait for this future here the guard will not work since while you are waiting someone else can initiate another config change. The guard works only if there is no preemption between checking and adding an entry.

Reply:
Yes, it requires a double-check, and maybe an additional semaphore. But with this patch, there is no way to do it any more if you want to avoid exception.

Reply:
I don't get what's wrong with waiting rather than throwing. raft::server can wait, it's not raft::fsm

Reply:
The purpose of this PR is not to change the API but to fix some bugs.
The existing API design was that we throw if there's an in-progress change.

If you want to change/improve the API it deserves a separate PR, it's out of scope of this one. Introducing a bunch of concurrency primitives into the code in this PR would miss the point.

Reply:
> A typical "guard" for issuing is get_configuration().is_joint(). It used to be a reliable way to find out if you can append a new config.

If you did that - appended a new config at the right moment, just after fsm appended non-joint entry but before server observed it - you would break the entire thing, that's the whole point of introducing this change - to fix that bug.

Reply:
This is a new bug, introduced by Petr's fix. You're appealing to status quo but what's the point of doing it if it's broken? Now you're saying that changing/improving the api is a separate fix - maybe, but this patch breaks a scenario that used to work. If you're fine breaking it - so be it, but it does look like we're layering patches on patches.

Reply:
Specifically, the patch is *adding* code that throws. I dared to ask, why not instead *add* the code that waits. Honestly, guys, you *own* it so if you feel OK owning a code that can throw - great. I find such code hard to reason about, it has no clear contract.

Reply:
Before this PR the code would throw. After this PR the code also throws. The time period during which it throws simply got extended. The contract is clear - if there is an in-progress configuration change, we throw. That was also the contract before. What changed is the definition of "in-progress". Previously the definition used `fsm`, leaking internal implementation details of `server` to the outside world - so we wouldn't consider a change in progress if `server` didn't finish it but a part of it (`fsm`) did. Now the `server` must finish the change before it allows the next one.

If you want to discuss this further, and change from throwing to waiting, or improve the interface otherwise (for example, we could have a `bool server::configuration_change_in_progress()` function) - I suggest we do it in a separate issue and any decisions go in a separate pull request.